executor/sample.go

```diff
@@ -0,0 +1,372 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"context"
+	"sort"
+	"time"
+
+	"github.com/cznic/mathutil"
+	"github.com/opentracing/opentracing-go"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser/model"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/kv"
+	plannercore "github.com/pingcap/tidb/planner/core"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/store/tikv"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	decoder "github.com/pingcap/tidb/util/rowDecoder"
+)
+
+var _ Executor = &TableSampleExecutor{}
+
+const sampleMethodRegionConcurrency = 5
+
+// TableSampleExecutor fetches a few rows through kv.Scan
+// according to the specific sample method.
+type TableSampleExecutor struct {
+	baseExecutor
+
+	table     table.Table
+	startTS   uint64
+	tablePlan plannercore.PhysicalPlan
+
+	sampler rowSampler
+}
+
+// Open initializes necessary variables for using this executor.
+func (e *TableSampleExecutor) Open(ctx context.Context) error {
+	if span := opentracing.SpanFromContext(ctx); span != nil && span.Tracer() != nil {
+		span1 := span.Tracer().StartSpan("TableSampleExecutor.Open", opentracing.ChildOf(span.Context()))
+		defer span1.Finish()
+	}
+	return nil
+}
+
+// Next fills data into the chunk passed by its caller.
+// The task was actually done by sampler.
+func (e *TableSampleExecutor) Next(ctx context.Context, req *chunk.Chunk) error {
+	req.Reset()
+	if e.sampler.finished() {
+		return nil
+	}
+	return e.sampler.writeChunk(req)
+}
+
+// Close implements the Executor Close interface.
+func (e *TableSampleExecutor) Close() error {
+	return nil
+}
+
+type rowSampler interface {
+	writeChunk(req *chunk.Chunk) error
+	finished() bool
+}
+
+type tableRegionSampler struct {
+	ctx        sessionctx.Context
+	table      table.Table
+	startTS    uint64
+	partTables []table.PartitionedTable
+	schema     *expression.Schema
+	fullSchema *expression.Schema
+	isDesc     bool
+	retTypes   []*types.FieldType
+
+	rowMap       map[int64]types.Datum
+	restKVRanges []kv.KeyRange
+	isFinished   bool
+}
+
+func newTableRegionSampler(ctx sessionctx.Context, t table.Table, startTs uint64, partTables []table.PartitionedTable,
+	schema *expression.Schema, fullSchema *expression.Schema, retTypes []*types.FieldType, desc bool) *tableRegionSampler {
```

Comment:
Should we need to consider refreshForUpdateTSForRC() ?

Reply:
I think it doesn't matter.